### PR TITLE
[Profile] Visualize the query plan and query profile

### LINKED ISSUE #5474: [Proposal] Visualize the query plan and query profile
**Is your feature request related to a problem? Please describe.**

Currently, Doris query plan and query profile are only be viewed in plain text format, which is not easy
to view and analyze the relationship of each node in plan or profile tree.

So I decide to visualize them to help user analyze them more easily.

**Describe the solution you'd like**

**Query Plan**

Query plan is a relative simple structure. Each plan fragment in the query plan is already a tree structure.
What we do is to connect all fragments together (DataSink -> ExchangeNode), and then draw a tree graph
for it.

```
                       ┌──────────────┐                       
                       │[2: DATA SINK]│                       
                       │Fragment: 2]  │                       
                       │RESULT SINK   │                       
                       └──────────────┘                       
                               │                              
                               │                              
              ┌─────────────────────────────────┐             
              │[2: HASH JOIN]                   │             
              │Fragment: 2]                     │             
              │join op: INNER JOIN (PARTITIONED)│             
              └─────────────────────────────────┘             
               ┌───────────────┴──────────────┐               
               │                              │               
        ┌─────────────┐                ┌─────────────┐        
        │[3: EXCHANGE]│                │[4: EXCHANGE]│        
        │Fragment: 2] │                │Fragment: 2] │        
        └─────────────┘                └─────────────┘        
               └┐                             └┐              
                │                              │              
 ┌────────────────────────────┐ ┌────────────────────────────┐
 │[3: DATA SINK]              │ │[4: DATA SINK]              │
 │Fragment: 0]                │ │Fragment: 1]                │
 │STREAM DATA SINK            │ │STREAM DATA SINK            │
 │  EXCHANGE ID: 03           │ │  EXCHANGE ID: 04           │
 │  HASH_PARTITIONED: `a`.`k1`│ │  HASH_PARTITIONED: `b`.`k1`│
 └────────────────────────────┘ └────────────────────────────┘
               ┌┘                             ┌┘              
               │                              │               
      ┌─────────────────┐            ┌─────────────────┐      
      │[0: OlapScanNode]│            │[1: OlapScanNode]│      
      │Fragment: 0]     │            │Fragment: 1]     │      
      │TABLE: tbl1      │            │TABLE: tbl1      │      
      └─────────────────┘            └─────────────────┘      
```

**Query Profile**

Query profile is more complicate because it does not only show the relationship of each plan fragments.
But it also shows all instances in each plan fragments.
So I plan to divide the presentation of the profile into several steps:

1. Fragment Tree

    The fragment tree is a tree used to display the overall query execution structure. They show the connection relationship 
    between each fragment, and the connection relationship of the execution nodes within each fragment.
    And at the same time, the execution time of the instance that takes the longest in each Fragment will be displayed.

    Through this figure, the user can understand the overall execution structure, and can further expand a fragment to
    continue analyzing the profile according to the maximum execution time.

```
              ┌──────────────────────┐
              │[-1: DataBufferSender]│
              │Fragment: 0           │
              │MaxActiveTime: 7.850ms│
              └──────────────────────┘
                          │
                          │
                ┌───────────────────┐
                │[2: HASH_JOIN_NODE]│
                │Fragment: 0        │
                └───────────────────┘
             ┌────────────┴─────────────┐
             │                          │
   ┌──────────────────┐       ┌──────────────────┐
   │[3: EXCHANGE_NODE]│       │[4: EXCHANGE_NODE]│
   │Fragment: 0       │       │Fragment: 0       │
   └──────────────────┘       └──────────────────┘
             │                          │
             │                          │
┌────────────────────────┐ ┌────────────────────────┐
│[3: DataStreamSender]   │ │[4: DataStreamSender]   │
│Fragment: 2             │ │Fragment: 1             │
│MaxActiveTime: 218.157us│ │MaxActiveTime: 562.931us│
└────────────────────────┘ └────────────────────────┘
            ┌┘                         ┌┘
            │                          │
  ┌───────────────────┐      ┌───────────────────┐
  │[0: OLAP_SCAN_NODE]│      │[1: OLAP_SCAN_NODE]│
  │Fragment: 2        │      │Fragment: 1        │
  └───────────────────┘      └───────────────────┘
            │                          │
            │                          │
     ┌─────────────┐            ┌─────────────┐
     │[OlapScanner]│            │[OlapScanner]│
     │Fragment: 2  │            │Fragment: 1  │
     └─────────────┘            └─────────────┘
            │                          │
            │                          │
   ┌─────────────────┐        ┌─────────────────┐
   │[SegmentIterator]│        │[SegmentIterator]│
   │Fragment: 2      │        │Fragment: 1      │
   └─────────────────┘        └─────────────────┘
```

2. Instance List

    The user can continue to expand a specified fragment, and then get a list of all instances of this fragment.
    The list contains instance id, instance execution host and instance execution time.
    Further, the user can continue to expand an instance for further analysis.

```
    +----------------------------------+------------------+------------+
    | Instances                        | Host             | ActiveTime |
    +----------------------------------+------------------+------------+
    | f8db008ccdf4ed8-807102c30d795eaa | 10.81.85.89:9062 | 7.850ms    |
    +----------------------------------+------------------+------------+
```

3. Instance Tree

    Instance Tree will display the detailed information of each execution node in an instance.
    Here is the most fine-grained information of the profile

```
                            ┌──────────────────────────────┐
                            │[-1: DataBufferSender]        │
                            │(Active: 0ns, non-child: 0.00)│
                            │  - Counters:                 │
                            │      - AppendBatchTime: 0ns  │
                            │      - NumSentRows: 0        │
                            └──────────────────────────────┘
                                            │
                                            │
                          ┌───────────────────────────────────┐
                          │[2: HASH_JOIN_NODE]                │
                          │(Active: 7.842ms, non-child: 0.00) │
                          │  - Counters:                      │
                          │      - BuildBuckets: 1.024K (1024)│
                          │      - BuildRows: 0               │
                          │      - BuildTime: 6.989us         │
                          │      - LoadFactor: 0.00           │
                          │      - PeakMemoryUsage: 48.00 KB  │
                          │      - ProbeRows: 0               │
                          │      - ProbeTime: 0ns             │
                          │      - PushDownComputeTime: 0ns   │
                          │      - PushDownTime: 0ns          │
                          │      - RowsReturned: 0            │
                          │      - RowsReturnedRate: 0        │
                          └───────────────────────────────────┘
                      ┌─────────────────────┴──────────────────────┐
                      │                                            │
┌──────────────────────────────────────────┐ ┌──────────────────────────────────────────┐
│[3: EXCHANGE_NODE]                        │ │[4: EXCHANGE_NODE]                        │
│(Active: 7.779ms, non-child: 23.54)       │ │(Active: 6.440ms, non-child: 19.49)       │
│  - Counters:                             │ │  - Counters:                             │
│      - BytesReceived: 0.00               │ │      - BytesReceived: 0.00               │
│      - ConvertRowBatchTime: 572ns        │ │      - ConvertRowBatchTime: 623ns        │
│      - DataArrivalWaitTime: 7.771ms      │ │      - DataArrivalWaitTime: 6.432ms      │
│      - DeserializeRowBatchTimer: 0ns     │ │      - DeserializeRowBatchTimer: 0ns     │
│      - FirstBatchArrivalWaitTime: 7.771ms│ │      - FirstBatchArrivalWaitTime: 6.431ms│
│      - PeakMemoryUsage: 0.00             │ │      - PeakMemoryUsage: 0.00             │
│      - RowsReturned: 0                   │ │      - RowsReturned: 0                   │
│      - RowsReturnedRate: 0               │ │      - RowsReturnedRate: 0               │
│      - SendersBlockedTotalTimer(*): 0ns  │ │      - SendersBlockedTotalTimer(*): 0ns  │
└──────────────────────────────────────────┘ └──────────────────────────────────────────┘
```

**Describe alternatives you've considered**

First, I will only implement this in text graph.
Next, It can be show in web frontend.

### PR DESCRIPTION
**Pull request by morningman**
## Proposed changes

Add command:
1. EXPLAIN GRAPGH SELECT ...
2. SHOW QUERY PROFILE "..."

Document will be added in next PR.
Details can be seem in #5474 

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Other

I introduced a tree printer class from https://github.com/davidsusu/tree-printer
This is a greet tree printer, but it release version has a NPE bug.
So I can't just add maven dependency of it, but has to add all its source codes in
our repo and fix the bug.

I will also send a PR to `davidsusu/tree-printer`, and once it merged and release a new version.
I will remove the source code of the tree printer in Doris and use its maven dependency.

Update:
Using tree-printer 1.2 instead of source code

## For Reviewer:

1.
The file in package: `fe/fe-common/src/main/java/org/apache/doris/common/treeprinter/`
are from `davidsusu/tree-printer`, no need to review.

2.
Add a new statement: `SHOW QUERY PROFILE "/"`;

3.
Files under `fe/fe-core/src/main/java/org/apache/doris/common/profile/` are for tree construction
and tree printer.

### CODE REVIEW

**Comment by EmmyMiao87** [comment]
Can't his tool directly reference jar? Do we have to buckle the code into our own code base?

**Comment by morningman** [comment]
> > Can't his tool directly reference jar? Do we have to buckle the code into our own code base?
> 
> OK, I will wait it to make to new release.

@EmmyMiao87 Done

File: fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.doris.planner.PlanNodeId;
+import org.apache.doris.thrift.TExplainLevel;
+
+import com.clearspring.analytics.util.Lists;
+
+import java.util.List;
+
+public class PlanTreeBuilder {
+
+    private static final String EXCHANGE_NODE = "EXCHANGE";
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java:33]: Use final params in Exchange Node.

File: fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    private static final String MERGING_EXCHANGE_NODE = "MERGING-EXCHANGE";
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java:34]: Same as above

File: fe/fe-core/src/main/cup/sql_parser.cup [hunk 3]
```diff
@@ -2653,14 +2661,14 @@ describe_stmt ::=
...
+        query.setIsExplain(options);
         RESULT = query;
     :}   
     | describe_command insert_stmt:stmt
     {:
-        stmt.getQueryStmt().setIsExplain(true);
+        stmt.getQueryStmt().setIsExplain(new ExplainOptions(true, false));
         RESULT = stmt;
     :}
     ;
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/cup/sql_parser.cup:2674]: KW_DESCRIBE means both "desc" and "explain",
see `fe/fe-core/src/main/jflex/sql_scanner.flex`

File: fe/fe-core/src/main/java/org/apache/doris/analysis/StatementBase.java [hunk 4]
```diff
@@ -36,9 +36,7 @@
     private String clusterName;
 
     // True if this QueryStmt is the top level query from an EXPLAIN <query>
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/StatementBase.java:38]: ok

File: fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        for (PlanTreeNode exchangeNode : exchangeNodes) {
+            if (exchangeNode.getId().equals(senderId)) {
+                return exchangeNode;
+            }
+        }
+        return null;
+    }
+
+    private void buildForPlanNode(PlanNode planNode, PlanTreeNode parent) {
+        StringBuilder sb = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java:107]: ok

File: fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    private void buildFragmentPlans() {
+        int i = 0;
+        for (PlanFragment fragment : fragments) {
+            DataSink sink = fragment.getSink();
+            PlanTreeNode sinkNode = null;
+            if (sink != null) {
+                StringBuilder sb = new StringBuilder();
+                sb.append("[").append(sink.getExchNodeId().asInt()).append(": DATA SINK]");
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java:61]: Change to sink's class name

File: fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java (hunk 1 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java:33]: Move this to ExchangeNode

File: fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java (hunk 2 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/common/profile/PlanTreeBuilder.java:34]: Done

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**